File: langs/csharp/src/VMx/Aggregates/AggregateVM4.cs
using VMx.Builders;
using VMx.Components;
using VMx.Services;

namespace VMx.Aggregates;

/// <summary>
/// Arity-4 aggregate viewmodel. A fixed tuple of four heterogeneous component VMs.
/// Component slots are populated lazily on construct() by invoking factories
/// provided via the nested builder.
///
/// See spec/08-aggregate-vm.md and ADR-0007.
/// </summary>
/// <typeparam name="VM1">Type of the first component.</typeparam>
/// <typeparam name="VM2">Type of the second component.</typeparam>
/// <typeparam name="VM3">Type of the third component.</typeparam>
/// <typeparam name="VM4">Type of the fourth component.</typeparam>
public sealed class AggregateVM4<VM1, VM2, VM3, VM4> : ComponentVMBase, IAggregateVM4<VM1, VM2, VM3, VM4>, IAggregateSlots
    where VM1 : class, IComponentVM
    where VM2 : class, IComponentVM
    where VM3 : class, IComponentVM
    where VM4 : class, IComponentVM
{
    private readonly Func<VM1> _factory1;
    private readonly Func<VM2> _factory2;
    private readonly Func<VM3> _factory3;
    private readonly Func<VM4> _factory4;
    private VM1? _component1;
    private VM2? _component2;
    private VM3? _component3;
    private VM4? _component4;

    IEnumerable<IComponentVM> IAggregateSlots.EnumerateSlots()
    {
        if (_component1 is { } c1) yield return c1;
        if (_component2 is { } c2) yield return c2;
        if (_component3 is { } c3) yield return c3;
        if (_component4 is { } c4) yield return c4;
    }

    // ── IAggregateVM4<VM1, VM2, VM3, VM4> ───────────────────────────────────

    /// <inheritdoc/>
    public VM1? Component1 => _component1;

    /// <inheritdoc/>
    public VM2? Component2 => _component2;

    /// <inheritdoc/>
    public VM3? Component3 => _component3;

    /// <inheritdoc/>
    public VM4? Component4 => _component4;

    // ── IComponentVM.Type ───────────────────────────────────────────────────

    /// <inheritdoc/>
    public override ViewModelType Type => ViewModelType.Aggregate;

    // ── Constructor ─────────────────────────────────────────────────────────

    private AggregateVM4(
        string name,
        string hint,
        IMessageHub hub,
        IDispatcher dispatcher,
        Func<VM1> factory1,
        Func<VM2> factory2,
        Func<VM3> factory3,
        Func<VM4> factory4)
        : base(name, hint, hub, dispatcher, onConstruct: null, onDestruct: null)
    {
        _factory1 = factory1;
        _factory2 = factory2;
        _factory3 = factory3;
        _factory4 = factory4;
    }

    // ── Lifecycle overrides ─────────────────────────────────────────────────

    /// <inheritdoc/>
    protected override void OnConstruct()
    {
        // On Reconstruct, dispose previous slot instances before overwriting
        // so their hub subscriptions and command Subjects don't leak.
        _component1?.Dispose();
        _component2?.Dispose();
        _component3?.Dispose();
        _component4?.Dispose();

        _component1 = _factory1();
        NotifyPropertyChanged(nameof(Component1));

        _component2 = _factory2();
        NotifyPropertyChanged(nameof(Component2));

        _component3 = _factory3();
        NotifyPropertyChanged(nameof(Component3));

        _component4 = _factory4();
        NotifyPropertyChanged(nameof(Component4));

        CompleteLifecycleHookAfter(TransitionChildrenAsync(
            [_component1, _component2, _component3, _component4],
            construct: true));
    }

    /// <inheritdoc/>
    protected override void OnDestruct()
    {
        CompleteLifecycleHookAfter(TransitionChildrenAsync(
            new IComponentVM?[] { _component1, _component2, _component3, _component4 }
                .OfType<IComponentVM>(),
            construct: false));
    }

    /// <summary>
    /// Dispose cascade (LIFE-013): dispose each component slot depth-first, then self.
    /// </summary>
    public override void Dispose()
    {
        var firstError = DisposeChildren([_component1, _component2, _component3, _component4]);
        try { base.Dispose(); }
        catch (Exception error)
        {
            firstError ??= System.Runtime.ExceptionServices.ExceptionDispatchInfo.Capture(error);
        }
        firstError?.Throw();
    }

    // ── Builder factory ─────────────────────────────────────────────────────

    /// <summary>Returns a new empty builder for <see cref="AggregateVM4{VM1,VM2,VM3,VM4}"/>.</summary>
    public static AggregateVM4Builder Builder() => new();

    // ── Nested builder ──────────────────────────────────────────────────────

    /// <summary>
    /// Immutable fluent builder for <see cref="AggregateVM4{VM1,VM2,VM3,VM4}"/>.
    /// Each setter returns a NEW builder instance (BLD-001).
    /// </summary>
    public sealed class AggregateVM4Builder
    {
        private readonly string? _name;
        private readonly string _hint;
        private readonly IMessageHub? _hub;
        private readonly IDispatcher? _dispatcher;
        private readonly Func<VM1>? _factory1;
        private readonly Func<VM2>? _factory2;
        private readonly Func<VM3>? _factory3;
        private readonly Func<VM4>? _factory4;

        internal AggregateVM4Builder()
        {
            _hint = "";
        }

        private AggregateVM4Builder(
            string? name,
            string hint,
            IMessageHub? hub,
            IDispatcher? dispatcher,
            Func<VM1>? factory1,
            Func<VM2>? factory2,
            Func<VM3>? factory3,
            Func<VM4>? factory4)
        {
            _name = name;
            _hint = hint;
            _hub = hub;
            _dispatcher = dispatcher;
            _factory1 = factory1;
            _factory2 = factory2;
            _factory3 = factory3;
            _factory4 = factory4;
        }

        /// <summary>Sets the required Name field.</summary>
        public AggregateVM4Builder Name(string name)
            => new(name, _hint, _hub, _dispatcher, _factory1, _factory2, _factory3, _factory4);

        /// <summary>Sets the optional Hint field (default: empty string).</summary>
        public AggregateVM4Builder Hint(string hint)
            => new(_name, hint, _hub, _dispatcher, _factory1, _factory2, _factory3, _factory4);

        /// <summary>Sets the required Services (hub + dispatcher).</summary>
        public AggregateVM4Builder Services(IMessageHub hub, IDispatcher dispatcher)
            => new(_name, _hint, hub, dispatcher, _factory1, _factory2, _factory3, _factory4);

        /// <summary>Sets the required factory for Component1.</summary>
        public AggregateVM4Builder Component1(Func<VM1> factory)
            => new(_name, _hint, _hub, _dispatcher, factory, _factory2, _factory3, _factory4);

        /// <summary>Sets the required factory for Component2.</summary>
        public AggregateVM4Builder Component2(Func<VM2> factory)
            => new(_name, _hint, _hub, _dispatcher, _factory1, factory, _factory3, _factory4);

        /// <summary>Sets the required factory for Component3.</summary>
        public AggregateVM4Builder Component3(Func<VM3> factory)
            => new(_name, _hint, _hub, _dispatcher, _factory1, _factory2, factory, _factory4);

        /// <summary>Sets the required factory for Component4.</summary>
        public AggregateVM4Builder Component4(Func<VM4> factory)
            => new(_name, _hint, _hub, _dispatcher, _factory1, _factory2, _factory3, factory);

        /// <summary>
        /// Validates required fields and constructs a <see cref="AggregateVM4{VM1,VM2,VM3,VM4}"/>.
        /// Throws <see cref="BuilderValidationException"/> if a required field is missing.
        /// </summary>
        public AggregateVM4<VM1, VM2, VM3, VM4> Build()
        {
            BuilderValidationException.Require(_name, "Name");
            BuilderValidationException.Require(_hub, "Hub");
            BuilderValidationException.Require(_dispatcher, "Dispatcher");
            BuilderValidationException.Require(_factory1, "Component1");
            BuilderValidationException.Require(_factory2, "Component2");
            BuilderValidationException.Require(_factory3, "Component3");
            BuilderValidationException.Require(_factory4, "Component4");

            return new AggregateVM4<VM1, VM2, VM3, VM4>(
                _name, _hint, _hub, _dispatcher,
                _factory1, _factory2, _factory3, _factory4);
        }
    }
}

File: langs/typescript/src/aggregates/aggregateVM4.ts
/**
 * AggregateVM4<VM1..VM4> — arity-4 aggregate viewmodel.
 *
 * See spec/08-aggregate-vm.md.
 */
import { ComponentVMBase } from "../components/componentVMBase.js";
import { ViewModelType } from "../components/types.js";
import type { IMessageHub } from "../services/messageHub.js";
import type { IDispatcher } from "../services/dispatcher.js";
import { BuilderValidationError } from "../builders/exceptions.js";
import { AggregateParent, commitAggregateSlots, validateAggregateSlots } from "./ownership.js";

const SENTINEL = Symbol("not-set");

export class AggregateVM4<
  VM1 extends ComponentVMBase,
  VM2 extends ComponentVMBase,
  VM3 extends ComponentVMBase,
  VM4 extends ComponentVMBase,
> extends ComponentVMBase {
  readonly #aggregateParent: AggregateParent;
  readonly #factory1: () => VM1;
  readonly #factory2: () => VM2;
  readonly #factory3: () => VM3;
  readonly #factory4: () => VM4;
  #component1: VM1 | null = null;
  #component2: VM2 | null = null;
  #component3: VM3 | null = null;
  #component4: VM4 | null = null;

  constructor(opts: {
    name: string; hint: string; hub: IMessageHub; dispatcher: IDispatcher;
    factory1: () => VM1; factory2: () => VM2; factory3: () => VM3; factory4: () => VM4;
  }) {
    super(opts);
    this.#aggregateParent = new AggregateParent(this, () => this.components());
    this.#factory1 = opts.factory1; this.#factory2 = opts.factory2;
    this.#factory3 = opts.factory3; this.#factory4 = opts.factory4;
  }

  get type(): ViewModelType { return ViewModelType.Aggregate; }
  get component1(): VM1 | null { return this.#component1; }
  get component2(): VM2 | null { return this.#component2; }
  get component3(): VM3 | null { return this.#component3; }
  get component4(): VM4 | null { return this.#component4; }

  /**
   * VMX-023: component slots in declaration order (null slots omitted). Tree
   * traversal uses this typed accessor instead of `component${i}` reflection.
   */
  components(): readonly ComponentVMBase[] {
    const slots: readonly (ComponentVMBase | null)[] = [
      this.#component1,
      this.#component2,
      this.#component3,
      this.#component4,
    ];
    return slots.filter((c): c is ComponentVMBase => c !== null);
  }

  protected override _onConstruct(): void {
    const next1 = this.#factory1();
    const next2 = this.#factory2();
    const next3 = this.#factory3();
    const next4 = this.#factory4();
    validateAggregateSlots(this.#aggregateParent, [next1, next2, next3, next4]);
    const previous = [this.#component1, this.#component2, this.#component3, this.#component4];
    // On Reconstruct, dispose previous slot instances before overwriting
    // so their hub subscriptions and command Subjects don't leak.
    this.#component1?.dispose();
    this.#component2?.dispose();
    this.#component3?.dispose();
    this.#component4?.dispose();

    this.#component1 = next1;
    this._notifyPropertyChanged("component1");

    this.#component2 = next2;
    this._notifyPropertyChanged("component2");

    this.#component3 = next3;
    this._notifyPropertyChanged("component3");

    this.#component4 = next4;
    commitAggregateSlots(this.#aggregateParent, previous, [next1, next2, next3, next4]);
    this._notifyPropertyChanged("component4");

    this.#component1.construct();
    this.#component2.construct();
    this.#component3.construct();
    this.#component4.construct();
  }

  protected override _onDestruct(): void {
    this.#component1?.destruct();
    this.#component2?.destruct();
    this.#component3?.destruct();
    this.#component4?.destruct();
  }

  override dispose(): void {
    // Depth-first dispose (LIFE-013): each component slot first, then self.
    this.#component1?.dispose();
    this.#component2?.dispose();
    this.#component3?.dispose();
    this.#component4?.dispose();
    super.dispose();
  }

  static builder<
    VM1 extends ComponentVMBase,
    VM2 extends ComponentVMBase,
    VM3 extends ComponentVMBase,
    VM4 extends ComponentVMBase,
  >(): AggregateVM4Builder<VM1, VM2, VM3, VM4> {
    return new AggregateVM4Builder<VM1, VM2, VM3, VM4>();
  }
}

export class AggregateVM4Builder<
  VM1 extends ComponentVMBase,
  VM2 extends ComponentVMBase,
  VM3 extends ComponentVMBase,
  VM4 extends ComponentVMBase,
> {
  #name: string | null = null;
  #hint = "";
  #hub: IMessageHub | null = null;
  #dispatcher: IDispatcher | null = null;
  #factory1: (() => VM1) | typeof SENTINEL = SENTINEL;
  #factory2: (() => VM2) | typeof SENTINEL = SENTINEL;
  #factory3: (() => VM3) | typeof SENTINEL = SENTINEL;
  #factory4: (() => VM4) | typeof SENTINEL = SENTINEL;

  constructor(from?: AggregateVM4Builder<VM1, VM2, VM3, VM4>) {
    if (from) {
      this.#name = from.#name; this.#hint = from.#hint;
      this.#hub = from.#hub; this.#dispatcher = from.#dispatcher;
      this.#factory1 = from.#factory1; this.#factory2 = from.#factory2;
      this.#factory3 = from.#factory3; this.#factory4 = from.#factory4;
    }
  }

  name(v: string): AggregateVM4Builder<VM1, VM2, VM3, VM4> { const b = new AggregateVM4Builder<VM1, VM2, VM3, VM4>(this); b.#name = v; return b; }
  hint(v: string): AggregateVM4Builder<VM1, VM2, VM3, VM4> { const b = new AggregateVM4Builder<VM1, VM2, VM3, VM4>(this); b.#hint = v; return b; }
  services(hub: IMessageHub, disp: IDispatcher): AggregateVM4Builder<VM1, VM2, VM3, VM4> { const b = new AggregateVM4Builder<VM1, VM2, VM3, VM4>(this); b.#hub = hub; b.#dispatcher = disp; return b; }
  component1(f: () => VM1): AggregateVM4Builder<VM1, VM2, VM3, VM4> { const b = new AggregateVM4Builder<VM1, VM2, VM3, VM4>(this); b.#factory1 = f; return b; }
  component2(f: () => VM2): AggregateVM4Builder<VM1, VM2, VM3, VM4> { const b = new AggregateVM4Builder<VM1, VM2, VM3, VM4>(this); b.#factory2 = f; return b; }
  component3(f: () => VM3): AggregateVM4Builder<VM1, VM2, VM3, VM4> { const b = new AggregateVM4Builder<VM1, VM2, VM3, VM4>(this); b.#factory3 = f; return b; }
  component4(f: () => VM4): AggregateVM4Builder<VM1, VM2, VM3, VM4> { const b = new AggregateVM4Builder<VM1, VM2, VM3, VM4>(this); b.#factory4 = f; return b; }

  build(): AggregateVM4<VM1, VM2, VM3, VM4> {
    if (this.#name === null) throw new BuilderValidationError("name");
    if (this.#hub === null || this.#dispatcher === null) throw new BuilderValidationError("services");
    if (this.#factory1 === SENTINEL) throw new BuilderValidationError("component1");
    if (this.#factory2 === SENTINEL) throw new BuilderValidationError("component2");
    if (this.#factory3 === SENTINEL) throw new BuilderValidationError("component3");
    if (this.#factory4 === SENTINEL) throw new BuilderValidationError("component4");
    const factory1 = this.#factory1;
    const factory2 = this.#factory2;
    const factory3 = this.#factory3;
    const factory4 = this.#factory4;
    return new AggregateVM4<VM1, VM2, VM3, VM4>({
      name: this.#name, hint: this.#hint, hub: this.#hub, dispatcher: this.#dispatcher,
      factory1, factory2, factory3, factory4,
    });
  }
}
